### ISSUE TITLE
Connection Pool Resiliency

### ISSUE THREAD

**Issue by mrapczynski** [oracle/node-oracledb#560]
I wonder about the behavior of the connection pool, and whether or not it has internal logic to replenish the pool in the face of intermittent connection failures. Occasionally due to our own issues, say for instance a blip in the network, a connection pool becomes exhausted and never replenishes the connections. Here is an example of log data I captured today:

<img width="898" alt="screen shot 2016-11-29 at 4 21 04 pm" src="https://cloud.githubusercontent.com/assets/258614/20734704/20c37e18-b650-11e6-80fc-709a589ae7b9.png">

In this scenario, eventually the request to `execute(...)` fails because the pool wouldn't hand out a connection due the reported timeout. But during that period, our database was never down or unavailable. As I suggested above, this is a moment in time where a connection failed or couldn't be made. But a restart of the app fixes it.

Is it reasonable to assume that `node-oracledb` self diagnoses and tries to eliminate bad connections, or is it less sophisticated than hoped?

**Comment by cjbj** [comment]
Once a connection gets a connection error and subsequently closes the connection, it will be dropped from the pool.  So a future `getConnection()`  that grows the pool should give you a new, valid connection.  See https://github.com/oracle/node-oracledb/blob/master/src/dpi/src/dpiConnImpl.cpp#L739 
Are you seeing this?  Make sure you are releasing connections in your error paths.  If you are, and you still see problems where you run out of connections, send us a test case.

If you link node-oracledb with Oracle Client 12.2 libraries (currently available on Oracle Cloud), the OCI layer has a lightweight, always enabled status check when dispensing sessions from the session pool (as used by node-oracledb's connection pool) which will likely help your situation. Each `getConnection()` will return a usable connection (though your app needs to cope if the network outage happens between `getConnection()` and `execute()`)

From the [12.2 OCI Manual](http://docs.oracle.com/database/122/LNOCI/release-changes.htm#LNOCI005) 

> "Improvements to OCI Session Pooling involving dead connection detection, cleanup, and idle session re-establishment. This improves reliability, giving applications more certainty that OCISessionGet() can return a usable session after an underlying network outage or database node failure."

For Oracle Client 11.2 & 12.1 users, in node-oracledb 1.12 (coming soon), @pvenkatraman has added the pool pinging feature mentioned in #300.  This is time based: the ping will happen if the connection hasn't been used in a configurable amount of time.

Bottom line: check you are releasing connections in all code paths, and wait a few days until the dev release of 1.12 is out.

**Comment by mrapczynski** [comment]
@cjbj This is an extraordinarily helpful response, thank you very much. I have a project in the hopper that I can use to road test the suggested improvements.

I don't always understand Oracle versioning so is the currently available 12.1.0.2.0 what you mean by 12.2?

**Comment by cjbj** [comment]
@mrapczynski I really did mean Oracle 12.2 which is just out, but currently only on [Oracle Database Cloud](https://cloud.oracle.com/database). The 'on premise' and Instant Client release is yet to come. The always-on check I mentioned only needs the new 'client side' 12.2 libraries, so once you have access to Oracle 12.2 libraries you should be able to play with that bit of the feature without upgrading your DB.  This is because Oracle client libraries of one version can connect to an Oracle DB of a different version. If you can't use the Oracle 12.2 client libraries, you can adjust the new `oracledb.poolPingInterval` property in node-oracledb 1.12 to meet your QoS and scalability requirements.

**Comment by cjbj** [comment]
@mrapczynski I just pushed node-oracledb 1.12.0-dev to GitHub (not npm).  Try it out.  See https://github.com/oracle/node-oracledb/blob/v1.12.0-dev/doc/api.md#connpoolpinging and https://github.com/oracle/node-oracledb/blob/v1.12.0-dev/doc/api.md#propdbpoolpinginterval

**Comment by mrapczynski** [comment]
@cjbj OK. I just installed it now, and I'll report back if I have any problems. Thanks!

**Comment by cjbj** [comment]
@mrapczynski any comments on 1.12.0-dev?

**Comment by mrapczynski** [comment]
@cjbj So far it's looking good, but more significant UAT should start around next week for the project I'm using as a test case so I'll keep you posted

**Comment by cjbj** [comment]
Looking forward to hearing how you tune `poolPingInterval`.

**Comment by mrapczynski** [comment]
@cjbj So far OK. I'm just sticking with the default ping interval, but it's also extraordinarily hard to predict when we have those network problems. It could very well be working nicely in the background (which is the goal anyway!).

Do you expose any kind of event emitter in the driver for diagnostics? Like if we wanted to do something like `Oracle.on('connectionPing', () => { ... });`, say for for example reporting pings or ping failures to our logging subsystem. I don't think I saw anything in the docs, but wanted to double-check.

Also question about pooling when the database is legitimately down for maintenance. A few times a year our DB is put into restricted mode so we can apply vendor patches to our ERP software. In those cases, all client connections are terminated. It seems like our Node applications are fine with this, and start re-opening new connections on demand as soon as the DB accepts them again. How do you think the pinging behavior might affected the ability of the pool to police itself in these instances?

In the Java world, we often had to restart apps after maintenance windows to get the connection pools alive again -- hence my curiosity with how things work with the Node driver.

**Comment by cjbj** [comment]
@mrapczynski you're right about the goal.  And no, there are no events being emitted.  I'm not currently thinking of adding any since when using a 12.2+ client the driver would never know when pinging occurs, since it is all handled at a lower level.  It is all supposed to be transparent.  
Pinging should aid your planned maintenance situation, but for best application behavior, you might also want to structure the application so it won't even try to connect at those times.

**Comment by cjbj** [comment]
@mrapczynski node-oracledb 1.12.2 is out on npm.  Let me know how 1.12 works for you.

**Comment by cjbj** [comment]
@mrapczynski ping

**Comment by mrapczynski** [comment]
@cjbj The pilot project I referred to in prior comments is now out to production with the latest driver. No issues, and everything seems to be working well. Time will tell when we have our next maintenance window if the pinging helps. I can close out this issue.

Any ETA on the next iteration of the Instant Client driver we talked about?

**Comment by cjbj** [comment]
@mrapczynski Good to hear.  Let me know how it all goes.

I can't give an ETA on 12.2 Instant Client.  Until it is out, only Oracle Cloud users have access to the updated Oracle libraries.

**Comment by mrapczynski** [comment]
@cjbj I think the pinging mechanism is working pretty good. I have been hunting some bugs in one of our apps, and as I roam the office, I have been haphazardly switching between VPN over Wi-Fi and ethernet cable depending on where I am. I can see a connection error or two when roaming, but then our app picks up right up and keeps going once my network connection is re-established.

Well done!

**Comment by cjbj** [comment]
Excellent.  And eventually when 12.2 is available those last remaining errors shouldn't occur.

Thanks for the feedback.  

Positive quotes about the driver always help us justify what we do; feel free to lavish praise!!

**Comment by cjbj** [comment]
@mrapczynski Instant Client 12.2 for Linux x64 is now out on http://www.oracle.com/technetwork/topics/linuxx86-64soft-092277.html
Other platforms will come.

**Comment by mrapczynski** [comment]
@cjbj Ha this is funny. I actually just checked Oracle yesterday and still not see it released. Thanks very much for the alert!

**Comment by cjbj** [comment]
@mrapczynski no problems.  Share the word!
